<html>
<head>
 <script type="text/javascript" src="/site_media/jquery/js/jquery-1.6.1.min.js"></script>

 <style type="text/css">
  #formwrap {
   line-height: 2em;
   background: #eef;
   margin: 10px;
   padding: 10px;
   height: 500px;
   text-align: center;
  }
  body {
   font-family: "Lucida Grande", "Lucida Sans Unicode", Verdana, Arial, Helvetica, sans-serif;
   font-size: 14px;
  }
  .help {cursor:help; border-bottom: 1px dotted #A9A9A9}
 </style>

 <script>
  $(function()
  {
    // turn off async JavaScript so we can easily get/set var names
    $.ajaxSetup({async: false});
    var descriptions = {};
    function addAdvancedSettings() {
    var libType = $("#librarytype option:selected").val();
    var varType = (document.getElementById('variationtype')[document.variantcaller.variationtype.selectedIndex].text).replace(/\s/g,'_');
    var paramsFileName = libType + "." + varType + ".json";
    $.get( '/pluginMedia/variantCaller/'+paramsFileName, function(data) {
	if(data) {
		var lines = data.match(/^.*((\r\n|\n|\r)|$)/gm);
		var currGroup = "";
		var commentText = "";
		$.each(lines, function(intIndex, objValue) {
				if(objValue.search("#") != -1) {
					objValue = objValue.replace(/\#/g, "");
					commentText = commentText + objValue + " ";
				}
				else if(objValue.search(":") != -1) {
                			var tokens = objValue.split(":");
					tokens[0] = tokens[0].replace(/\s/g, "");
					tokens[1] = tokens[1].replace(/\s/g, "");
				
					if(tokens[1] == "{") {
						currGroup = tokens[0];
						commentText = "";
					}
					else {
						var labelText = tokens[0];
						var paramName = labelText.replace(/-/g, "___");
						var paramValue = tokens[1].replace(/,$/, "");
						if($("#"+paramName).length > 0) {
							$("#"+paramName).val(paramValue);
						}
						else {
							//var desc = commentText;
							var desc = "";
							if(desc == "") {
								desc = descriptions[labelText]; 
							}
							var htmlText = "<tr>";
							var labelName = "label_"+paramName;
							htmlText = htmlText + "<td><span class='help' id=" + labelName +"> </span> </td>";
							htmlText = htmlText + "<td><input type='text' size='10' id=" + paramName + "> </td> </tr>";
							$("#"+currGroup).append(htmlText);
							$("#"+paramName).val(paramValue);
							var groupName = currGroup.replace(/-/g, "___");
							$("#"+paramName).attr("name", "param__" + groupName +"__" + paramName);
							$("#"+labelName).text(labelText+":");
							$("#"+labelName).attr("title", desc);
						}
						commentText = "";
					}
				}
		});
        }
    });
    }

    function loadDescriptions() {
    var libType = "ampliseq";
    var varType = "Germ_Line";
    var paramsFileName = libType + "." + varType + ".json";
    $.get( '/pluginMedia/variantCaller/'+paramsFileName, function(data) {
	if(data) {
		var lines = data.match(/^.*((\r\n|\n|\r)|$)/gm);
		var currGroup = "";
		var commentText = "";
		$.each(lines, function(intIndex, objValue) {
				if(objValue.search("#") != -1) {
					objValue = objValue.replace(/\#/g, "");
					commentText = commentText + objValue + " ";
				}
				else if(objValue.search(":") != -1) {
                			var tokens = objValue.split(":");
					tokens[0] = tokens[0].replace(/\s/g, "");
					tokens[1] = tokens[1].replace(/\s/g, "");
				
					if(tokens[1] == "{") {
						currGroup = tokens[0];
						commentText = "";
					}
					else {
						var paramName = tokens[0];
						var paramValue = tokens[1].replace(/,$/, "");
						descriptions[paramName] = commentText;
					}
					commentText = "";
				}
		});
        }
    });
    }
    
    loadDescriptions();
    addAdvancedSettings();

    $("#librarytype").change(function() {
      var opt = $("#librarytype option:selected").val();
      if( opt == "ampliseq" ) {
        $("#ptrim").show();
        $("#tpadding").hide();
        $("#ustarts").hide();
      } else if( opt == "targetseq" ) {
        $("#ptrim").hide();
        $("#tpadding").show();
        $("#ustarts").hide();
      } else {
        $("#ptrim").hide();
        $("#tpadding").hide();
        $("#ustarts").hide();
      }
      addAdvancedSettings();
    });

    $("#variationtype").change(function() {
    	addAdvancedSettings();
    });
 });

function toggle2(id, id2, link) {
        var e = document.getElementById(id);
        if (e.style.display == '') {
                e.style.display = 'none';
                link.innerHTML = 'Frequently Asked Questions         +';
        } 
        else {
                e.style.display = '';
                link.innerHTML = 'Frequently Asked Questions         -';
        }
        document.getElementById(id2).style.display = 'none';
        document.getElementById('advText').innerHTML = 'Advanced Settings         +';
  }

  function toggle(id, id2, link) {
        var e = document.getElementById(id);
        if (e.style.display == '') {
                e.style.display = 'none';
                link.innerHTML = 'Advanced Settings         +';
        } 
        else {
                e.style.display = '';
                link.innerHTML = 'Advanced Settings         -';
        }
        document.getElementById(id2).style.display = 'none';
        document.getElementById('faqText').innerHTML = 'Frequently Asked Questions         +';
 }
 </script>
</head>

<body>
 <div align="center">
  <form id="variantcaller" name="variantcaller" style="margin-left:auto;margin-right:auto">
   <h1>Torrent Variant Caller Plugin</h1>
   <table cellpadding="5px">
    <tr>
     <td><span class="help" title="Select the library (enrichment) type sequenced.">Library Type:</span></td>
     <td><select id="librarytype" name="librarytype" style="width:120px">
      <option value="fullgenome">Full Genome</option>
      <option value="ampliseq">Ion AmpliSeq</option>
      <option value="targetseq">Ion TargetSeq</option>
     </select></td>
    </tr>
    <tr>
     <td><span class="help" title="Select the type of variants you are interested in detecting.">Variant Frequency:</span></td>
     <td><select id="variationtype" name="variationtype" style="width:120px">
      <option value="Germ_Line">Germ Line</option>
      <option value="Somatic">Somatic</option>
     </select></td>
    </tr>
    <tr id="ptrim" style="display:none">
    <td><span class="help"
       title="Trim reads to amplicon targets. Primarily to avoid variant sites from being covered by residual primers from overlapping amplicons.">
       Trim Reads:</span></td>
     <td><input type="checkbox" id="trimreads" name="trimreads" checked="checked" value="Yes"/></td>
    </tr>
    <tr id="tpadding" style="display:none">
     <td><span class="help"
       title="Number of bases target regions are extended at both ends. A value greater than 0 allows variants to be detected in regions where individual reads overlap the boundaries of targeted regions.">
       Target Padding:</span></td>
     <td><input type="text" size=6 id="padtargets" name="padtargets" value=0></select></td>
    </tr>
    <tr id="ustarts" style="display:none">
     <td><span class="help"
       title="Analyze variants for a sampling of on-target reads using just one read starting at each reference position for both read orientations. This removes bias due to non-uniform library enrichment but produces a most conservative representation of target coverage.">
       Use Unique Starts:</span></td>
     <td><input type="checkbox" id="uniquestarts" name="uniquestarts" value="No"/></td> 
    </tr>
   </table>
   <br/>
   <br/>
   <p> <H3> <a href="#" id='advText' onClick="toggle('advanced', 'faq', this)"> Advanced Settings                +</a> </H3> </p>
   <p> <H3> <a href="#" id='faqText' onClick="toggle2('faq', 'advanced', this)"> Frequently Asked Questions                +</a> </H3> </p>

   <div id="faq" width="60%" style="display:none">

   <table bgcolor="#EFF5FB" width="60%" >
   <tr><td>
 <b> 1) What is new in this version of the Torrent Variant Caller (TVC) ? </b>
                <p>In this version of the plugin you will be able to adjust all the parameters used by the Variant Caller and set the same parameters when creating a  “Plan”. For information about creating a plan please refer to the “Torrent Browser User Interface Guide” .</p>
  </td></tr>
  <tr><td>
    <b>2) Will my parameters be saved for the next run when I am using the TVC interface?</b>
        <p> No.  Every time you run the plugin the default parameters will be displayed, you can change all parameters, but they will not be saved.  It is advisable to take note of the changes you make. This information will be useful in case of needing troubleshooting.</p>
  </td></tr>
  <tr> <td> 
  <b>3) How can I use my parameters repeatedly without typing them over and over?</b>
        <p>You can create a “Plan” of analysis that can be saved and applied to several runs. Plan creation allows setting all the parameters for the TVC. For information about creating a plan please refer to the “Torrent Browser User Interface Guide” </p>
   </td></tr>
   <tr><td>
    <b>4) Are there any performance improvements in this new version of TVC?</b>
        <p> The calling of INDELs is slightly faster.
        There is an  increase  in the number of True Positives and a  decrease in the number of False Positives for INDEL calling.</p>
    </td></tr>
    <tr><td>
<b> 5) Where can I get more information about the TVC? </b>
        <br>
        <a href="http://ioncommunity.lifetechnologies.com/login.jspa" target="_blank">Overview of Variant Calling in Ion Sequencing Data </a> <br>
        <a href="http://ioncommunity.lifetechnologies.com/login.jspa" target="_blank">Torrent Variant Caller Plugin </a> <br>
        <a href="http://ioncommunity.lifetechnologies.com/login.jspa" target="_blank">Advanced Topics in Variant Calling </a> <br>
   </td></tr>
   </table>
   </div>

   <div id="advanced" style="display:none">
   <table border="2" bgcolor="#EFF5FB">
        <tr>
        <td width="49%"> 
        <div id="dibayes_advanced">
        <p> <H4> <u> SNP Caller Settings: </u> </H4> </p>
        <table id="dibayes"> 
        </table>
        </div>
        </td>
        <td width="2%">
        </td>
        <td width="49%">
        <div id="tvc_advanced">
        <p> <H4> <u> Indel Caller Settings: </u> </H4> </p>
        <table id="torrent-variant-caller"> 
        </table>
        </div>
        <div id="assembly_advanced">
        <p> <H4> <u> Long Indel Assembly Settings: </u> </H4> </p>
        <table id="long-indel-assembler"> </table>
        </div>
        <div id="filter_advanced">
        <p><H4> <u> Indel Filter Advanced Settings: </u> </H4> </p>
        <table id="filter-indels"> </table>
        </div>
        </td>
        </tr>
   </table>
   </div>
   <div align="left">
    <h3>Description and Usage Notes</h3>
    <p>
     This plugin analyzes mapped reads coverage over each individual reference base to deduce whether there is sufficient
     variation to call a SNP or INDEL at that position. If a target region and/or hotspot region is defined, then this
     analysis is restricted to just those regions and more detailed information is presented on the report pages.</p>
    <p>
     The target regions selected should correspond to your reference and enriched fragment library. You may also select
     'Whole genome' to analyze coverage against the full reference, but this option is only recommended if your library (type)
     is for a small genome and is not enriched.</p> 
    <p>
     A run fails almost immediately (after queuing) if your selected targeted regions do not appear to match
     the reference that the reads were mapped to. (Check the log file if the plugin status is 'Error').</p>
    <p>
     NOTE: This plugin is designed to work with alignment files containing Ion Torrent flowspace information and will fail
     to call variants for older alignments without flow order information or for combined alignments generated from runs
     that use differrent flow sequences.</p>
   </div>
  </form>
  </div>
</body>
</html>

