\hypertarget{struct_align_stats_1_1options}{
\section{AlignStats::options Struct Reference}
\label{struct_align_stats_1_1options}\index{AlignStats::options@{AlignStats::options}}
}


{\ttfamily \#include $<$alignStats.h$>$}

\subsection*{Public Attributes}
\begin{DoxyCompactItemize}
\item 
string \hyperlink{struct_align_stats_1_1options_ad85a9162387d560e42421b81b138de99}{bam\_\-file}
\item 
string \hyperlink{struct_align_stats_1_1options_ac837b3aaf6b013cd76c8b90916e507f7}{bam\_\-index}
\item 
string \hyperlink{struct_align_stats_1_1options_aeccd6951719549cde348e6cd58e43283}{out\_\-file}
\item 
string \hyperlink{struct_align_stats_1_1options_ab115525327af943f92ad59b6e00b7747}{genome\_\-info}
\item 
long \hyperlink{struct_align_stats_1_1options_a533ba263882d33023b63682a5de8aa62}{filter\_\-length}
\item 
string \hyperlink{struct_align_stats_1_1options_af5705e123a1b8dc2fb5414415c6311bb}{q\_\-scores}
\item 
long \hyperlink{struct_align_stats_1_1options_a90fbedd6612bb1635a9a078619d03bdd}{start\_\-slop}
\item 
int \hyperlink{struct_align_stats_1_1options_a37bd731bc14bf63365db2a82122b0609}{sam\_\-parsed\_\-flag}
\item 
long \hyperlink{struct_align_stats_1_1options_a13b8f29a0ad066ba915ee118c21c9b8d}{total\_\-reads}
\item 
long \hyperlink{struct_align_stats_1_1options_ab168ff2db5170e036226d3938e993321}{sample\_\-size}
\item 
bool \hyperlink{struct_align_stats_1_1options_ae9a1d04b390a861656a13bbb9becefcc}{help\_\-flag}
\item 
bool \hyperlink{struct_align_stats_1_1options_a83d2e3d53aacd64b8f2cef6f09ae4d2e}{skip\_\-cov\_\-flag}
\item 
bool \hyperlink{struct_align_stats_1_1options_ab1d57cdf5b105d06aa184d41af5e4505}{debug\_\-flag}
\item 
int \hyperlink{struct_align_stats_1_1options_a5e25dc83b0fc3c8e955bf4267c6dc7d2}{num\_\-threads}
\item 
int \hyperlink{struct_align_stats_1_1options_aadebe419d97d862d6441d1563367e246}{buffer\_\-size}
\item 
bool \hyperlink{struct_align_stats_1_1options_aed2b228a0a13d98c5d544a48d4091b09}{iupac\_\-flag}
\item 
bool \hyperlink{struct_align_stats_1_1options_a64eb45282d8d26af88352817d84cd94a}{keep\_\-iupac}
\item 
int \hyperlink{struct_align_stats_1_1options_af7fdf5340fa114c3296fd3f675b48b4f}{max\_\-coverage}
\item 
string \hyperlink{struct_align_stats_1_1options_a5b6f3a50319bd5bea781818bd6c83aa0}{align\_\-summary\_\-file}
\item 
int \hyperlink{struct_align_stats_1_1options_a4477e10888176b696fbe80caec742fc2}{align\_\-summary\_\-min\_\-len}
\item 
int \hyperlink{struct_align_stats_1_1options_a119003c51a782cc237af14b164d529a3}{align\_\-summary\_\-max\_\-len}
\item 
int \hyperlink{struct_align_stats_1_1options_a2e2cd52d35bdc24f73ad1c61e8638ab0}{align\_\-summary\_\-len\_\-step}
\item 
int \hyperlink{struct_align_stats_1_1options_a2d920885b7f36b06c6bd08d2634cabef}{align\_\-summary\_\-max\_\-errors}
\item 
int \hyperlink{struct_align_stats_1_1options_ae9989ba9b255f20000d48ac1920bba31}{align\_\-summary\_\-filter\_\-len}
\item 
double \hyperlink{struct_align_stats_1_1options_ac96a618a5c1c0f097c7e3f0c92c4d709}{align\_\-summary\_\-filter\_\-accuracy}
\item 
bool \hyperlink{struct_align_stats_1_1options_ac71f2469bdd8a1303cb7fbc21bf6d61a}{stdin\_\-sam\_\-flag}
\item 
bool \hyperlink{struct_align_stats_1_1options_a134d6b9c72e42947b581cc9d88c7f219}{stdin\_\-bam\_\-flag}
\item 
std::string \hyperlink{struct_align_stats_1_1options_a9e16b499aaad5e645652bbc234dc018f}{list\_\-of\_\-files}
\item 
bool \hyperlink{struct_align_stats_1_1options_a1df99156054317dbb94f18912ee045b7}{truncate\_\-soft\_\-clipped}
\end{DoxyCompactItemize}


\subsection{Detailed Description}
A structure to keep the command line options for an \hyperlink{class_align_stats}{AlignStats} instance This structure is used to keep the constructors very simple. 

\subsection{Member Data Documentation}
\hypertarget{struct_align_stats_1_1options_a5b6f3a50319bd5bea781818bd6c83aa0}{
\index{AlignStats::options@{AlignStats::options}!align\_\-summary\_\-file@{align\_\-summary\_\-file}}
\index{align\_\-summary\_\-file@{align\_\-summary\_\-file}!AlignStats::options@{AlignStats::options}}
\subsubsection[{align\_\-summary\_\-file}]{\setlength{\rightskip}{0pt plus 5cm}string {\bf AlignStats::options::align\_\-summary\_\-file}}}
\label{struct_align_stats_1_1options_a5b6f3a50319bd5bea781818bd6c83aa0}
a string representing the absolute path for the alignment summary error table \hypertarget{struct_align_stats_1_1options_ac96a618a5c1c0f097c7e3f0c92c4d709}{
\index{AlignStats::options@{AlignStats::options}!align\_\-summary\_\-filter\_\-accuracy@{align\_\-summary\_\-filter\_\-accuracy}}
\index{align\_\-summary\_\-filter\_\-accuracy@{align\_\-summary\_\-filter\_\-accuracy}!AlignStats::options@{AlignStats::options}}
\subsubsection[{align\_\-summary\_\-filter\_\-accuracy}]{\setlength{\rightskip}{0pt plus 5cm}double {\bf AlignStats::options::align\_\-summary\_\-filter\_\-accuracy}}}
\label{struct_align_stats_1_1options_ac96a618a5c1c0f097c7e3f0c92c4d709}
a double representing the \% of the read that must be error free in order to be considered for the alignment summary error table. This filter value doesn't effect the alignment.summary or sam.parsed output. Filtering formula: if (((align\_\-summary\_\-filter\_\-len -\/ cummaltive\_\-errors\_\-in\_\-read) / align\_\-summary\_\-filter\_\-len) $>$ align\_\-summary\_\-filter\_\-accuracy) \hypertarget{struct_align_stats_1_1options_ae9989ba9b255f20000d48ac1920bba31}{
\index{AlignStats::options@{AlignStats::options}!align\_\-summary\_\-filter\_\-len@{align\_\-summary\_\-filter\_\-len}}
\index{align\_\-summary\_\-filter\_\-len@{align\_\-summary\_\-filter\_\-len}!AlignStats::options@{AlignStats::options}}
\subsubsection[{align\_\-summary\_\-filter\_\-len}]{\setlength{\rightskip}{0pt plus 5cm}int {\bf AlignStats::options::align\_\-summary\_\-filter\_\-len}}}
\label{struct_align_stats_1_1options_ae9989ba9b255f20000d48ac1920bba31}
an int represting the minimum lenght a read must be in order to be considered for the alignment summary error table. This length is also used for the alignment.summary file and sam.parsed \hypertarget{struct_align_stats_1_1options_a2e2cd52d35bdc24f73ad1c61e8638ab0}{
\index{AlignStats::options@{AlignStats::options}!align\_\-summary\_\-len\_\-step@{align\_\-summary\_\-len\_\-step}}
\index{align\_\-summary\_\-len\_\-step@{align\_\-summary\_\-len\_\-step}!AlignStats::options@{AlignStats::options}}
\subsubsection[{align\_\-summary\_\-len\_\-step}]{\setlength{\rightskip}{0pt plus 5cm}int {\bf AlignStats::options::align\_\-summary\_\-len\_\-step}}}
\label{struct_align_stats_1_1options_a2e2cd52d35bdc24f73ad1c61e8638ab0}
an int representing the interval of read lengths between the min and max lengths in the alignment summary table \hypertarget{struct_align_stats_1_1options_a2d920885b7f36b06c6bd08d2634cabef}{
\index{AlignStats::options@{AlignStats::options}!align\_\-summary\_\-max\_\-errors@{align\_\-summary\_\-max\_\-errors}}
\index{align\_\-summary\_\-max\_\-errors@{align\_\-summary\_\-max\_\-errors}!AlignStats::options@{AlignStats::options}}
\subsubsection[{align\_\-summary\_\-max\_\-errors}]{\setlength{\rightskip}{0pt plus 5cm}int {\bf AlignStats::options::align\_\-summary\_\-max\_\-errors}}}
\label{struct_align_stats_1_1options_a2d920885b7f36b06c6bd08d2634cabef}
an int representing the maximum number of errors to be output to the table. one column is produced for each +1 value until the value stored in align\_\-summary\_\-max\_\-errors. For instance, if this is 3 the columns output would be: err1 err2 err3+ The final column always has a trailing + character. This indicates that reads in this column have at least that number of errors \hypertarget{struct_align_stats_1_1options_a119003c51a782cc237af14b164d529a3}{
\index{AlignStats::options@{AlignStats::options}!align\_\-summary\_\-max\_\-len@{align\_\-summary\_\-max\_\-len}}
\index{align\_\-summary\_\-max\_\-len@{align\_\-summary\_\-max\_\-len}!AlignStats::options@{AlignStats::options}}
\subsubsection[{align\_\-summary\_\-max\_\-len}]{\setlength{\rightskip}{0pt plus 5cm}int {\bf AlignStats::options::align\_\-summary\_\-max\_\-len}}}
\label{struct_align_stats_1_1options_a119003c51a782cc237af14b164d529a3}
an int representing the maximum length for a read to be considered in the alignment summary error table \hypertarget{struct_align_stats_1_1options_a4477e10888176b696fbe80caec742fc2}{
\index{AlignStats::options@{AlignStats::options}!align\_\-summary\_\-min\_\-len@{align\_\-summary\_\-min\_\-len}}
\index{align\_\-summary\_\-min\_\-len@{align\_\-summary\_\-min\_\-len}!AlignStats::options@{AlignStats::options}}
\subsubsection[{align\_\-summary\_\-min\_\-len}]{\setlength{\rightskip}{0pt plus 5cm}int {\bf AlignStats::options::align\_\-summary\_\-min\_\-len}}}
\label{struct_align_stats_1_1options_a4477e10888176b696fbe80caec742fc2}
an int representing the minimum length for a read to be considered in the alignment summary error table \hypertarget{struct_align_stats_1_1options_ad85a9162387d560e42421b81b138de99}{
\index{AlignStats::options@{AlignStats::options}!bam\_\-file@{bam\_\-file}}
\index{bam\_\-file@{bam\_\-file}!AlignStats::options@{AlignStats::options}}
\subsubsection[{bam\_\-file}]{\setlength{\rightskip}{0pt plus 5cm}string {\bf AlignStats::options::bam\_\-file}}}
\label{struct_align_stats_1_1options_ad85a9162387d560e42421b81b138de99}
a string representing the absolute path to a bam file, or just a file name \hypertarget{struct_align_stats_1_1options_ac837b3aaf6b013cd76c8b90916e507f7}{
\index{AlignStats::options@{AlignStats::options}!bam\_\-index@{bam\_\-index}}
\index{bam\_\-index@{bam\_\-index}!AlignStats::options@{AlignStats::options}}
\subsubsection[{bam\_\-index}]{\setlength{\rightskip}{0pt plus 5cm}string {\bf AlignStats::options::bam\_\-index}}}
\label{struct_align_stats_1_1options_ac837b3aaf6b013cd76c8b90916e507f7}
a string representing the absolute path to a bam file index \hypertarget{struct_align_stats_1_1options_aadebe419d97d862d6441d1563367e246}{
\index{AlignStats::options@{AlignStats::options}!buffer\_\-size@{buffer\_\-size}}
\index{buffer\_\-size@{buffer\_\-size}!AlignStats::options@{AlignStats::options}}
\subsubsection[{buffer\_\-size}]{\setlength{\rightskip}{0pt plus 5cm}int {\bf AlignStats::options::buffer\_\-size}}}
\label{struct_align_stats_1_1options_aadebe419d97d862d6441d1563367e246}
an integer representing the number of reads that is assigned to each thread for a work load. If coverage is being calculated the buffer size is only approximately this size. It can be slightly larger, or smaller \hypertarget{struct_align_stats_1_1options_ab1d57cdf5b105d06aa184d41af5e4505}{
\index{AlignStats::options@{AlignStats::options}!debug\_\-flag@{debug\_\-flag}}
\index{debug\_\-flag@{debug\_\-flag}!AlignStats::options@{AlignStats::options}}
\subsubsection[{debug\_\-flag}]{\setlength{\rightskip}{0pt plus 5cm}bool {\bf AlignStats::options::debug\_\-flag}}}
\label{struct_align_stats_1_1options_ab1d57cdf5b105d06aa184d41af5e4505}
a bool that will print debug messages to stderr. very verbose, be careful \hypertarget{struct_align_stats_1_1options_a533ba263882d33023b63682a5de8aa62}{
\index{AlignStats::options@{AlignStats::options}!filter\_\-length@{filter\_\-length}}
\index{filter\_\-length@{filter\_\-length}!AlignStats::options@{AlignStats::options}}
\subsubsection[{filter\_\-length}]{\setlength{\rightskip}{0pt plus 5cm}long {\bf AlignStats::options::filter\_\-length}}}
\label{struct_align_stats_1_1options_a533ba263882d33023b63682a5de8aa62}
the minimum length a read must be to be included in calculations and output \mbox{[}Default: 20\mbox{]} \hypertarget{struct_align_stats_1_1options_ab115525327af943f92ad59b6e00b7747}{
\index{AlignStats::options@{AlignStats::options}!genome\_\-info@{genome\_\-info}}
\index{genome\_\-info@{genome\_\-info}!AlignStats::options@{AlignStats::options}}
\subsubsection[{genome\_\-info}]{\setlength{\rightskip}{0pt plus 5cm}string {\bf AlignStats::options::genome\_\-info}}}
\label{struct_align_stats_1_1options_ab115525327af943f92ad59b6e00b7747}
a string representing the absolute path to a genome.info.txt file (internal Ion Torrent file format \mbox{[}Default: \char`\"{}\char`\"{}\mbox{]} \hypertarget{struct_align_stats_1_1options_ae9a1d04b390a861656a13bbb9becefcc}{
\index{AlignStats::options@{AlignStats::options}!help\_\-flag@{help\_\-flag}}
\index{help\_\-flag@{help\_\-flag}!AlignStats::options@{AlignStats::options}}
\subsubsection[{help\_\-flag}]{\setlength{\rightskip}{0pt plus 5cm}bool {\bf AlignStats::options::help\_\-flag}}}
\label{struct_align_stats_1_1options_ae9a1d04b390a861656a13bbb9becefcc}
a bool to print the help message to stdout. exits the program if true \hypertarget{struct_align_stats_1_1options_aed2b228a0a13d98c5d544a48d4091b09}{
\index{AlignStats::options@{AlignStats::options}!iupac\_\-flag@{iupac\_\-flag}}
\index{iupac\_\-flag@{iupac\_\-flag}!AlignStats::options@{AlignStats::options}}
\subsubsection[{iupac\_\-flag}]{\setlength{\rightskip}{0pt plus 5cm}bool {\bf AlignStats::options::iupac\_\-flag}}}
\label{struct_align_stats_1_1options_aed2b228a0a13d98c5d544a48d4091b09}
a bool flag that will check the \hyperlink{class_m_d}{MD} tag for IUPAC codes and see if they're actually errors to the reference. use this if your reference contains IUPAC codes instead of bases in some positions \hypertarget{struct_align_stats_1_1options_a64eb45282d8d26af88352817d84cd94a}{
\index{AlignStats::options@{AlignStats::options}!keep\_\-iupac@{keep\_\-iupac}}
\index{keep\_\-iupac@{keep\_\-iupac}!AlignStats::options@{AlignStats::options}}
\subsubsection[{keep\_\-iupac}]{\setlength{\rightskip}{0pt plus 5cm}bool {\bf AlignStats::options::keep\_\-iupac}}}
\label{struct_align_stats_1_1options_a64eb45282d8d26af88352817d84cd94a}
a bool flag that will put iupac codes into the sam.parsed file. only has an effect if iupac\_\-flag == true \hypertarget{struct_align_stats_1_1options_a9e16b499aaad5e645652bbc234dc018f}{
\index{AlignStats::options@{AlignStats::options}!list\_\-of\_\-files@{list\_\-of\_\-files}}
\index{list\_\-of\_\-files@{list\_\-of\_\-files}!AlignStats::options@{AlignStats::options}}
\subsubsection[{list\_\-of\_\-files}]{\setlength{\rightskip}{0pt plus 5cm}std::string {\bf AlignStats::options::list\_\-of\_\-files}}}
\label{struct_align_stats_1_1options_a9e16b499aaad5e645652bbc234dc018f}
a string representing the absolute path to a file \hypertarget{struct_align_stats_1_1options_af7fdf5340fa114c3296fd3f675b48b4f}{
\index{AlignStats::options@{AlignStats::options}!max\_\-coverage@{max\_\-coverage}}
\index{max\_\-coverage@{max\_\-coverage}!AlignStats::options@{AlignStats::options}}
\subsubsection[{max\_\-coverage}]{\setlength{\rightskip}{0pt plus 5cm}int {\bf AlignStats::options::max\_\-coverage}}}
\label{struct_align_stats_1_1options_af7fdf5340fa114c3296fd3f675b48b4f}
an int representing the maximum coverage that is allowed for 1 position in the genome if a particular position has $>$ max\_\-coverage reads aligning to it, only the first N reads are used, where N = max\_\-coverage. \mbox{[}Default: 20000\mbox{]} \hypertarget{struct_align_stats_1_1options_a5e25dc83b0fc3c8e955bf4267c6dc7d2}{
\index{AlignStats::options@{AlignStats::options}!num\_\-threads@{num\_\-threads}}
\index{num\_\-threads@{num\_\-threads}!AlignStats::options@{AlignStats::options}}
\subsubsection[{num\_\-threads}]{\setlength{\rightskip}{0pt plus 5cm}int {\bf AlignStats::options::num\_\-threads}}}
\label{struct_align_stats_1_1options_a5e25dc83b0fc3c8e955bf4267c6dc7d2}
an integer representing the number of threads the program will try to use \hypertarget{struct_align_stats_1_1options_aeccd6951719549cde348e6cd58e43283}{
\index{AlignStats::options@{AlignStats::options}!out\_\-file@{out\_\-file}}
\index{out\_\-file@{out\_\-file}!AlignStats::options@{AlignStats::options}}
\subsubsection[{out\_\-file}]{\setlength{\rightskip}{0pt plus 5cm}string {\bf AlignStats::options::out\_\-file}}}
\label{struct_align_stats_1_1options_aeccd6951719549cde348e6cd58e43283}
a string representing the absolute path to the prefix of output file names \mbox{[}Default: \char`\"{}Default\char`\"{}\mbox{]} \hypertarget{struct_align_stats_1_1options_af5705e123a1b8dc2fb5414415c6311bb}{
\index{AlignStats::options@{AlignStats::options}!q\_\-scores@{q\_\-scores}}
\index{q\_\-scores@{q\_\-scores}!AlignStats::options@{AlignStats::options}}
\subsubsection[{q\_\-scores}]{\setlength{\rightskip}{0pt plus 5cm}string {\bf AlignStats::options::q\_\-scores}}}
\label{struct_align_stats_1_1options_af5705e123a1b8dc2fb5414415c6311bb}
a comma seperate string of integers reprsenting the phred values to use for the alignment.summary output \hypertarget{struct_align_stats_1_1options_a37bd731bc14bf63365db2a82122b0609}{
\index{AlignStats::options@{AlignStats::options}!sam\_\-parsed\_\-flag@{sam\_\-parsed\_\-flag}}
\index{sam\_\-parsed\_\-flag@{sam\_\-parsed\_\-flag}!AlignStats::options@{AlignStats::options}}
\subsubsection[{sam\_\-parsed\_\-flag}]{\setlength{\rightskip}{0pt plus 5cm}int {\bf AlignStats::options::sam\_\-parsed\_\-flag}}}
\label{struct_align_stats_1_1options_a37bd731bc14bf63365db2a82122b0609}
a flag to signal whether or not to create the sam.parsed file (can be very large file) \hypertarget{struct_align_stats_1_1options_ab168ff2db5170e036226d3938e993321}{
\index{AlignStats::options@{AlignStats::options}!sample\_\-size@{sample\_\-size}}
\index{sample\_\-size@{sample\_\-size}!AlignStats::options@{AlignStats::options}}
\subsubsection[{sample\_\-size}]{\setlength{\rightskip}{0pt plus 5cm}long {\bf AlignStats::options::sample\_\-size}}}
\label{struct_align_stats_1_1options_ab168ff2db5170e036226d3938e993321}
a long representing the number of reads to sample \hypertarget{struct_align_stats_1_1options_a83d2e3d53aacd64b8f2cef6f09ae4d2e}{
\index{AlignStats::options@{AlignStats::options}!skip\_\-cov\_\-flag@{skip\_\-cov\_\-flag}}
\index{skip\_\-cov\_\-flag@{skip\_\-cov\_\-flag}!AlignStats::options@{AlignStats::options}}
\subsubsection[{skip\_\-cov\_\-flag}]{\setlength{\rightskip}{0pt plus 5cm}bool {\bf AlignStats::options::skip\_\-cov\_\-flag}}}
\label{struct_align_stats_1_1options_a83d2e3d53aacd64b8f2cef6f09ae4d2e}
a bool that allows the user to skip coverage calculations. this speeds up the program considerably, and reduces memory use \hypertarget{struct_align_stats_1_1options_a90fbedd6612bb1635a9a078619d03bdd}{
\index{AlignStats::options@{AlignStats::options}!start\_\-slop@{start\_\-slop}}
\index{start\_\-slop@{start\_\-slop}!AlignStats::options@{AlignStats::options}}
\subsubsection[{start\_\-slop}]{\setlength{\rightskip}{0pt plus 5cm}long {\bf AlignStats::options::start\_\-slop}}}
\label{struct_align_stats_1_1options_a90fbedd6612bb1635a9a078619d03bdd}
an integer representing the first N bases to be ignored when considering errors in an alignment \hypertarget{struct_align_stats_1_1options_a134d6b9c72e42947b581cc9d88c7f219}{
\index{AlignStats::options@{AlignStats::options}!stdin\_\-bam\_\-flag@{stdin\_\-bam\_\-flag}}
\index{stdin\_\-bam\_\-flag@{stdin\_\-bam\_\-flag}!AlignStats::options@{AlignStats::options}}
\subsubsection[{stdin\_\-bam\_\-flag}]{\setlength{\rightskip}{0pt plus 5cm}bool {\bf AlignStats::options::stdin\_\-bam\_\-flag}}}
\label{struct_align_stats_1_1options_a134d6b9c72e42947b581cc9d88c7f219}
a bool representing whether or not the input is from stdin, and binary input \hypertarget{struct_align_stats_1_1options_ac71f2469bdd8a1303cb7fbc21bf6d61a}{
\index{AlignStats::options@{AlignStats::options}!stdin\_\-sam\_\-flag@{stdin\_\-sam\_\-flag}}
\index{stdin\_\-sam\_\-flag@{stdin\_\-sam\_\-flag}!AlignStats::options@{AlignStats::options}}
\subsubsection[{stdin\_\-sam\_\-flag}]{\setlength{\rightskip}{0pt plus 5cm}bool {\bf AlignStats::options::stdin\_\-sam\_\-flag}}}
\label{struct_align_stats_1_1options_ac71f2469bdd8a1303cb7fbc21bf6d61a}
a bool representing whether or not the input is from stdin, and raw text input \hypertarget{struct_align_stats_1_1options_a13b8f29a0ad066ba915ee118c21c9b8d}{
\index{AlignStats::options@{AlignStats::options}!total\_\-reads@{total\_\-reads}}
\index{total\_\-reads@{total\_\-reads}!AlignStats::options@{AlignStats::options}}
\subsubsection[{total\_\-reads}]{\setlength{\rightskip}{0pt plus 5cm}long {\bf AlignStats::options::total\_\-reads}}}
\label{struct_align_stats_1_1options_a13b8f29a0ad066ba915ee118c21c9b8d}
a long to represent the total number of reads being input. this is used for sampling purposes in order to extrapolate stats \hypertarget{struct_align_stats_1_1options_a1df99156054317dbb94f18912ee045b7}{
\index{AlignStats::options@{AlignStats::options}!truncate\_\-soft\_\-clipped@{truncate\_\-soft\_\-clipped}}
\index{truncate\_\-soft\_\-clipped@{truncate\_\-soft\_\-clipped}!AlignStats::options@{AlignStats::options}}
\subsubsection[{truncate\_\-soft\_\-clipped}]{\setlength{\rightskip}{0pt plus 5cm}bool {\bf AlignStats::options::truncate\_\-soft\_\-clipped}}}
\label{struct_align_stats_1_1options_a1df99156054317dbb94f18912ee045b7}
a bool to represent whether or not to ignore soft clipped bases in the alignment. This can really effect reads which map to the reverse strand and have a \# of leading soft clipped bases in the alignment \mbox{[}Default: true\mbox{]} 

The documentation for this struct was generated from the following file:\begin{DoxyCompactItemize}
\item 
alignStats.h\end{DoxyCompactItemize}
