<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html>
<html>
<head>
<script type="text/javascript" src="/site_media/jquery/js/jquery-1.6.1.min.js"></script>

<style type='text/css'> 
  #formwrap {
    line-height: 2em;
    background: #eef;
    margin: 10px;
    padding: 10px;
    height: 130px;
    text-align: center;
  }
  body {
    font-family: "Lucida Grande", "Lucida Sans Unicode", Verdana, Arial, Helvetica, sans-serif;
    font-size: 14px;
  }
  .center { margin-left:auto; margin-right:auto; }
  .help {cursor:help; border-bottom: 1px dotted #A9A9A9}
</style>

<script type='text/javascript'> 
$(function() {
  // disable development options
  var refFilter = "&short_name__in=hg19,mm10";    // use "" to view all
  $(".dev_only").hide();  // change to show() to see development options
  
  // get the reference for the current report and set as default for selector
  $.ajaxSetup({async: false});
  var reportUrl = '/rundb/api/v1/results/' + TB_result  + '/?format=json&noplugin=True';
  var referenceID;
  $.get( reportUrl, function(data) {
    referenceID = data.reference;
  });
  // get list of available references and selecion for current genome
  var seln = 0;
  $.ajax({
    url: "/rundb/api/v1/referencegenome/?limit=0&enabled=true&index_version=" + TB_TMAP + "&format=jsonp" + refFilter,
    success: function(data) {
      $("#reference").html("");
      if (data.meta.limit != 0) {
        $("#reference").html("<p> there are no plugins enabled </p>");
      }
      var items = "<option value=''>Select</option>";
      jQuery.each(data.objects,
      function(i, element) {
        refpath = element.reference_path + "/" + element.short_name + ".fasta";
        items += "<option value='" + refpath + "'>" + element.short_name + "</option>";
        if( element.short_name == referenceID ) seln = i+1;
      });
      $("#reference").html(items);
    },
    type: "GET",
    dataType: "jsonp"
  });
  $("#reference option")[seln].selected = true;

  $.fn.serializeObject = function() {
    var o = {};
    var a = this.serializeArray();
    $.each(a, function() {
      if (o[this.name] != null) {
        if (!o[this.name].push) {
          o[this.name] = [o[this.name]];
        }
        o[this.name].push(this.value || '');
      } else if(this.name == "file") {
        o[this.name] = [this.value || ''];  
      } else {
        o[this.name] = this.value || '';
      }
    });
    return o;
  };

  $(function() {
    $('#postbutton').click(function() {
      var refn = document.rnaseqform.reference.selectedIndex;
      if( refn == 0 ) {
        alert("Please select the Reference Genome");
        return false;
      }
      document.rnaseqform.genome.value = document.getElementById('reference')[refn].text;
      obj = $('#rnaseqform').serializeObject();
      pluginAPIJSON = { "plugin" : [TB_plugin.fields.name], "pluginconfig" : obj };
      pluginAPIJSON = JSON.stringify(pluginAPIJSON);
      pluginURL = "/rundb/api/v1/results/" + TB_result + "/plugin/";
      $.ajax({
        type: 'POST',
        url: pluginURL,
        contentType: "application/json; charset=utf-8",
        success: function(data) {
          $("#json_result").html( '<div style="text-align: center;">'+
            '<img src="/site_media/jquery/colorbox/images/loading.gif" alt="Running Plugin" style="float:center">'+
            '</img><p>Running the Plugin... Check RNASeqAnalysis.html in the Default Report </p></div>' );
        setTimeout("parent.$.fn.colorbox.close()",2000); },
        data: pluginAPIJSON,
        dataType: "json"
      });
    });
  });

});
</script>
</head>

<body>
 <form id="rnaseqform" name="rnaseqform">
  <div style="text-align:center">
   <h1>Torrent RNASeqAnalysis Plugin</h1>
   <h3>RNA Transcript Alignment and Analysis</h3>
  </div>
  <table class="center" cellpadding="5px">
   <tr>
    <td><span class="help" title="The short name of the reference genome for read alignment. Only a limited number of references are currently available for this plugin.">Reference Genome:</span></td>
    <td><select name="reference" id="reference" ></select></td> </tr>
   <tr class="dev_only" style="display:none">
    <td><span class="help" title="Choose a fraction of the reads to align. Use less than 100% to reduce analysis time.">Sample Reads:</span></td>
    <td><select name="fraction_of_reads">
     <option value="1" selected="selected">100%</option>
     <option value="0.5">50%</option>
     <option value="0.25">25%</option>
     <option value="0.1">10%</option>
     <option value="0.05">5%</option>
     <option value="0.01">1%</option>
     <option value="0.001">0.1%</option></select>
    </td> </tr>
   <tr class="dev_only" style="display:none">
    <td><span class="help" title="Specify the 3' adapter DNA sequence that might be seen towards the 3' end of reads. Usually the standard Ion Torrent adapter (or that specified in the run plan)  will be already trimmed off of Ion Torrent reads so the default value of 'None' is sufficient.">Adapter Sequence:</span></td>
    <td><input name="cutadapt" id="cutadapt" value="None"></td> </tr>
  </table>
  <br/>
  <div align="center" id="json_result">
   <input id="postbutton" type="submit" value="Submit">
  </div>
  <input type="hidden" id="genome" name="genome"/>
  <input type="hidden" id="launch_mode" name="launch_mode" value="Manual"/>
  <br/>
  <div align="left">
   <h3>Description and Usage Notes</h3>
   <p>
Use this plugin to analyze cDNA reads, as produced by RNA-Seq. Reads are aligned to the reference genome using STAR and bowtie2 aligners to find full and partial mappings. The alignments are analyzed by Picard tools to collect read counts and cufflinks to extract gene isoform representation. For barcoded data, comparative representation plots across barcodes are created in addition to individual reports for each barcode. All alignment, detailed and summary report files are made available for download.
   </p>
  </div>
 </form>   
</body>
</html>

