<html>
<head>
 <script type="text/javascript" src="/site_media/jquery/js/jquery-1.6.1.min.js"></script>

 <style type="text/css">
  #formwrap {
   line-height: 2em;
   background: #eef;
   margin: 10px;
   padding: 10px;
   height: 500px;
   text-align: center;
  }
  body {
   font-family: "Lucida Grande", "Lucida Sans Unicode", Verdana, Arial, Helvetica, sans-serif;
   font-size: 14px;
  }
  .help {cursor:help; border-bottom: 1px dotted #A9A9A9}
 </style>

 <script>
  $(function()
  {
    // turn off async JavaScript so we can easily get/set var names
    $.ajaxSetup({async: false});
    // get the reference for the current report - TS_result is defined in run verison of this script
    var reportUrl = '/rundb/api/v1/results/' + TB_result  + '/?format=json';
    var referenceID;
    var pluginOutDir;
    $.get( reportUrl, function(data)
    {
     var rl = data.log;
     var reportParams = rl.replace("log.html","ion_params_00.json");
     pluginOutDir = data.filesystempath + '/plugin_out/';
     $.getJSON( reportParams, function(data)
     {
      // from report data get the short reference name
      referenceID = data.libraryName;
     });
    });
    document.getElementById('referenceid').innerHTML = referenceID;

    function addBamSelection(dir) {
      var getBamListCmd = '/pluginMedia/coverageAnalysis/getTargets.php?ext=bam&dir=';
      $.getJSON( getBamListCmd+dir, function(data) {
        $.each( data, function( intIndex, objValue ) {
          var j = objValue.lastIndexOf('.bam')
          if( j > 0 )
          {
            var targfile = dir+'/'+objValue
            var targ = objValue.substring(0,j)
            $("#mergedbam").append("<option value='" + targfile + "'>" +  targ + "</option>")
          }
        });
      });
    }

    addBamSelection(pluginOutDir+'combineBarcodes_out');
    addBamSelection(pluginOutDir+'combineAlignments_out');
    addBamSelection('/results/imported/alignment/'+referenceID);

    var refBedURL = '/rundb/api/v1/content/?format=json&limit=0&publisher__name=BED&path__startswith=/' + referenceID + '/merged/plain/';
    $.get( refBedURL, function(data)
    {
      $.each( data.objects, function(intIndex,result)
      {
       var targfile = result.file;
       var i = targfile.lastIndexOf('/merged/plain/');
       if( i < 0 ) return true;
       var selName = targfile.substr(i+14)
       if( selName.substr(0,7) == 'example' ) return true;
       var j = selName.lastIndexOf('.bed');
       if( j > 0 && !result.meta.hotspot )
       {
        selName = selName.substr(0,j);
        $("#unptargets").append("<option value='" + targfile + "'>" +  selName + "</option>");
       }
      });
    });

    $.fn.serializeObject = function()
    {
     var o = {};
     var a = this.serializeArray();
     $.each(a,
      function() {
       if (o[this.name] != null) {
        if (!o[this.name].push) {
         o[this.name] = [o[this.name]];
        }
        o[this.name].push(this.value || '');
       } else {
        o[this.name] = this.value || '';
       }
      });
     return o;
    };

    $(function()
    {
     $('#postbutton').click(function()
     {
      var tarn = document.coverage.unptargets.selectedIndex;
      var regex = /^[0-9]+$/;
      var padval = document.coverage.padtargets.value;
      if( !regex.test(padval) )
      {
       alert("Target padding value must be a non-negative integer (>=0).");
       return false;
      }
      if( tarn == 0 && padval > 0 )
      {
       alert("Target padding value must be 0 if no specific Targeted regions value is selected.");
       return false;
      }
      var libn = document.coverage.librarytype.selectedIndex;
      var libType = document.getElementById('librarytype')[libn].value;
      var libName = document.getElementById('librarytype')[libn].text;
      if( libType == "fullgenome" )
      {
       if( tarn > 0 )
       {
        var msg = "Selecting targeted regions is typically not appropriate for Library Type: " + libName;
        msg += ". (Coverage will be considered in these just regions.)";
        msg += "\n\nDo you wish to continue?";
        if( !confirm(msg) ) return false;
       }
      }
      else
      {
       if( tarn == 0 )
       {
        var msg = "Targeted regions selection is typically appropriate for Library Type: " + libName;
        msg += ". (Variants may be reported in non-enriched regions where coverage is low.)";
        msg += "\n\nDo you wish to continue?";
        if( !confirm(msg) ) return false;
       }
      }
      var nondups = document.coverage.nonduplicates.checked;
      if( libType == "ampliseq" && nondups == true )
      {
       var msg = "The Non-duplicate reads option will be ignored for Library Type: " + libName;
       msg += ".\n\nDo you wish to continue?";
       if( !confirm(msg) ) return false;
      }
      // pass user selection with spaces translated to underscores
      document.coverage.mergedbam_id.value = document.getElementById('mergedbam')[document.coverage.mergedbam.selectedIndex].text;
      var usrSel = document.getElementById('librarytype')[libn].text;
      document.coverage.librarytype_id.value = usrSel.replace(/\s/g,'_');
      usrSel = document.getElementById('unptargets')[tarn].text;
      document.coverage.targetregions_id.value = usrSel.replace(/\s/g,'_');
      obj = $('#coverage').serializeObject();
      pluginAPIJSON = { "plugin" : [TB_plugin.fields.name], "pluginconfig" : obj };
      pluginAPIJSON = JSON.stringify(pluginAPIJSON);
      pluginURL = "/rundb/api/v1/results/" + TB_result + "/plugin/";
      $.ajax({
       type: 'POST',
       url: pluginURL,
       async: false,
       contentType: "application/json; charset=utf-8",
       success: function(data)
       {
        $("#json_result").html('<div style="text-align: center;"><img src="/site_media/jquery/colorbox/images/loading.gif" alt="Running Plugin" style="float:center"></img><p>Running the Plugin... Check coverageAnalysis.html in the Plugin Summary</p></div>');
        setTimeout("parent.$.fn.colorbox.close()",2000);
       },
       data: pluginAPIJSON,
       dataType: "json"
      });
     });
    });

  });
 </script>
</head>

<body>
 <div align="center">
  <form id="coverage" name="coverage" align="center">
   <h1>Torrent Coverage Analysis Plugin</h1>
   <h3>Genome and Targeted Re-sequencing Coverage Analysis</h3>
   <table cellpadding="5px">
    <tr>
     <td><span class="help" title="The short name of the reference genome (or DNA sequences) that the current report was generated against and to be used for coverage analysis.">Reference Genome:</span></td>
     <td><div id="referenceid"></div></td>
    </tr>
    <tr>
     <td><span class="help"
       title="Select the reads alignments you wish to analyze. If the combineAlignments plugin has been run in this report then the resulting combined aligmnents will be listed and available for selection. Otherwise the only option is 'Current Report'.">
       Aligned Reads:</span></td>
     <td><select id="mergedbam" name="mergedbam"><option value="">Current Report</option></select></td> </tr>
    <tr>
    <tr>
     <td><span class="help" title="Select the library (enrichment) type sequenced.">Library Type:</span></td>
     <td><select id="librarytype" name="librarytype">
      <option value="fullgenome">Full Genome</option>
      <option value="ampliseq">Ion AmpliSeq</option>
      <option value="targetseq">Ion TargetSeq</option>
     </select></td>
    </tr>
    <tr>
    <tr>
     <td><span class="help"
       title="Select the target regions (ROI) matching your reference and enriched fragment library. See Description and Notes below.">
       Targeted regions:</span></td>
     <td><select id="unptargets" name="unptargets"><option value="">None</option></select></td> </tr>
    <tr>
     <td><span class="help"
       title="Specifiy the number of bases for on/off padded targets analysis. See Description and Notes below.">
       Target padding (bases):</span></td>
     <td><input type="text" size=6 id="padtargets" name="padtargets" value=0></select></td> </tr>
    <tr>
     <td><span class="help"
       title="Check to analyze only reads mapped to unique positions in the genome. See Description and Usage Notes below.">
       Uniquely mapped reads:</span></td>
     <td><input type="checkbox" id="uniquemaps" name="uniquemaps" value="No" checked="checked"/></td> </tr>
    <tr>
     <td><span class="help"
       title="Check to analyze only reads that are not demarked as duplicate reads. See Description and Usage Notes below.">
       Non-duplicate reads:</span></td>
     <td><input type="checkbox" id="nonduplicates" name="nonduplicates" value="No" checked="checked"/></td> </tr>
   </table>
   <br/>
   <div id="json_result">
    <input id="postbutton" type="submit" value="Submit">
   </div>
   <input type="hidden" id="librarytype_id" name="librarytype_id"/>
   <input type="hidden" id="mergedbam_id" name="mergedbam_id"/>
   <input type="hidden" id="targetregions_id" name="targetregions_id"/>
   <br/>
   <div align="left">
    <h3>Description and Usage Notes</h3>
    <p>
     This plugin generates statistics, downloadable data and basic visualization of coverage over targeted regions of a
     reference.</p>
    <p>
     The targeted regions selected should correspond to your reference and enriched fragment library. Coverage will only
     be reported over the specified target region.</p>
    <p>
     The Target padding value specifies a number of bases to expand each individual target by at both 3' and 5' ends
     for analyzing padded target coverage. The recommended value to use for Ion TargetSeq is 100.
     All statistics and plots will be created using a padded set of (merged) targets.
     The padding value is ignorred if the selected library type is Full Genome.</p>
    <p>
     The 'Uniquely Mapped Reads' and 'Non-duplicate Reads' options are used to specify coverage analysis on a filtered
     subset of reads. Non-uniquely mapped reads are those that could have equally well aligned to multiple locations
     in the whole genome and are typically mapped randomly to one of these locations. Hence, such reads are typically
     no use for unambiguously locating variants and may result in a lower estimation of coverage for enriched libraries.
     For whole genome libraries these reads may still be an accurate estimation of coverage over repeated regions.
     Non-duplicate reads are those not marked as duplicates in the aligned reads (BAM) file and therefore the option
     to use these is only effective if duplicate reads were so marked during the alignment process.
    <p>
     A run will fail almost immediately (after queuing) if your selected targeted regions do not appear to match
     the reference that the reads were mapped to. (Check the log file if the plugin status is 'Error').</p>
   </div>
  </form>
  </div>
</body>
</html>

