<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE HTML>
<html>
<head>
 <script type="text/javascript" src="/site_media/jquery/js/jquery-1.6.1.min.js"></script>

 <style type="text/css">
  #formwrap {
   line-height: 2em;
   background: #eef;
   margin: 10px;
   padding: 10px;
   height: 500px;
  }
  body {
   font-family: "Lucida Grande", "Lucida Sans Unicode", Verdana, Arial, Helvetica, sans-serif;
   font-size: 14px;
  }
  .center { margin-left:auto; margin-right:auto; }
  .help {cursor:help; border-bottom: 1px dotted #A9A9A9}
 </style>

 <script>
  $(function()
  {
    // turn off async JavaScript so we can easily get/set var names
    $.ajaxSetup({async: false});
    // get the reference for the current report - TS_result is defined in run verison of this script
    var reportUrl = '/rundb/api/v1/results/' + TB_result  + '/?format=json';
    var referenceID;
    var pluginOutDir;
    $.get( reportUrl, function(data)
    {
     pluginOutDir = data.filesystempath + '/plugin_out/';
     referenceID = data.reference;
    });
    document.getElementById('referenceid').innerHTML = referenceID;

    var refBedType = '/unmerged/detail/';
    var refBedURL = '/rundb/api/v1/content/?format=json&limit=0&publisher__name=BED&path__startswith=/'+referenceID+refBedType;
    $.get( refBedURL, function(data)
    {
      $.each( data.objects, function(intIndex,result)
      {
       var targfile = result.file;
       var i = targfile.lastIndexOf(refBedType);
       if( i < 0 ) return true;
       var selName = targfile.substr(i+refBedType.length)
       if( selName.substr(0,7) == 'example' ) return true;
       var j = selName.lastIndexOf('.bed');
       if( j > 0 && !result.meta.hotspot )
       {
        selName = selName.substr(0,j);
        $("#targetregions").append("<option value='" + targfile + "'>" +  selName + "</option>");
       }
      });
    });

    $("#librarytype").change(function() {
      var opt = $("#librarytype option:selected").val();
      if( opt == "ampliseq" ) {
        $("#ptrim").show();
        $("#tpadding").hide();
        $("#uniquereads").show();
        $("#nondups").hide();
      } else if( opt == "targetseq" ) {
        $("#ptrim").hide();
        $("#tpadding").show();
        $("#uniquereads").show();
        $("#nondups").show();
      } else {
        $("#ptrim").hide();
        $("#tpadding").show();
        $("#uniquereads").show();
        $("#nondups").show();
      }
    });

    $.fn.serializeObject = function()
    {
     var o = {};
     var a = this.serializeArray();
     $.each(a,
      function() {
       if (o[this.name] != null) {
        if (!o[this.name].push) {
         o[this.name] = [o[this.name]];
        }
        o[this.name].push(this.value || '');
       } else {
        o[this.name] = this.value || '';
       }
      });
     return o;
    };

    $(function()
    {
     $('#postbutton').click(function()
     {
      var libn = document.coverageAnalysis.librarytype.selectedIndex;
      var libType = document.getElementById('librarytype')[libn].value;
      var libName = document.getElementById('librarytype')[libn].text;
      var tarn = document.coverageAnalysis.targetregions.selectedIndex;
      // in case padding set for a different type
      if( libType == "ampliseq" )
      {
        document.coverageAnalysis.padtargets.value = 0;
      }
      var regex = /^[0-9]+$/;
      var padval = document.coverageAnalysis.padtargets.value;
      if( !regex.test(padval) )
      {
       alert("Target padding value must be a non-negative integer (>=0).");
       return false;
      }
      if( tarn == 0 && padval > 0 )
      {
       alert("Target padding value must be 0 if no specific Targeted Regions value is selected.");
       return false;
      }
      if( libType == "wholegenome" )
      {
       if( tarn > 0 )
       {
        var msg = "Selecting targeted regions is typically not appropriate for Library Type: " + libName;
        msg += ". (Coverage will be considered in these just regions.)";
        msg += "\n\nDo you wish to continue?";
        if( !confirm(msg) ) return false;
       }
      }
      else
      {
       if( tarn == 0 )
       {
        var msg = "Targeted Regions selection is typically appropriate for Library Type: " + libName;
        msg += ". (Coverage may be reported in non-enriched regions where reads are not expected.)";
        msg += "\n\nDo you wish to continue?";
        if( !confirm(msg) ) return false;
       }
      }
      // force invisible options to be unchecked
      if( libType == "wholegenome" ) {
       $("#trimreads").attr("checked",false);
       $("#padtargets").attr("value","0");
      } else if( libType == "ampliseq" ) {
       $("#nonduplicates").attr("checked",false);
       $("#padtargets").attr("value","0");
      } else if( libType == "targetseq" ) {
       $("#trimreads").attr("checked",false);
      }
      // pass user selection with spaces translated to underscores
      var usrSel = document.getElementById('librarytype')[libn].text;
      document.coverageAnalysis.librarytype_id.value = usrSel.replace(/\s/g,'_');
      usrSel = document.getElementById('targetregions')[tarn].text;
      document.coverageAnalysis.targetregions_id.value = usrSel.replace(/\s/g,'_');
      obj = $('#coverageAnalysis').serializeObject();
      pluginAPIJSON = { "plugin" : [TB_plugin.fields.name], "pluginconfig" : obj };
      pluginAPIJSON = JSON.stringify(pluginAPIJSON);
      pluginURL = "/rundb/api/v1/results/" + TB_result + "/plugin/";
      $.ajax({
       type: 'POST',
       url: pluginURL,
       async: false,
       contentType: "application/json; charset=utf-8",
       success: function(data)
       {
        $("#json_result").html('<div style="text-align: center;"><img src="/site_media/jquery/colorbox/images/loading.gif" alt="Running Plugin" style="float:center"></img><p>Running the Plugin... Check coverageAnalysis.html in the Plugin Summary</p></div>');
        setTimeout("parent.$.fn.colorbox.close()",2000);
       },
       data: pluginAPIJSON,
       dataType: "json"
      });
     });
    });

  });
 </script>
</head>

<body>
 <form id="coverageAnalysis" name="coverageAnalysis">
  <div style="text-align:center">
   <h1>Torrent Coverage Analysis Plugin</h1>
   <h3>Genome and Targeted Re-sequencing Coverage Analysis</h3>
  </div>
   <table class="center" cellpadding="5px">
    <tr>
     <td><span class="help" title="The short name of the reference genome (or DNA sequences) that the current report was generated against and to be used for coverage analysis.">Reference Genome:</span></td>
     <td><div id="referenceid"></div></td>
    </tr>
    <tr>
     <td><span class="help" title="Select the library (enrichment) type sequenced.">Library Type:</span></td>
     <td><select id="librarytype" name="librarytype">
      <option value="wholegenome">Whole Genome</option>
      <option value="ampliseq" selected="selected">Ion AmpliSeq</option>
      <option value="targetseq">Ion TargetSeq</option>
     </select></td>
    </tr>
    <tr>
     <td><span class="help"
       title="Select the target regions (ROI) matching your reference and enriched fragment library. See Description and Usage Notes below.">
       Targeted Regions:</span></td>
     <td><select id="targetregions" name="targetregions"><option value="">None</option></select></td>
    </tr>
    <tr id="ptrim">
     <td><span class="help"
       title="Trim reads to amplicon targets. See Description and Usage Notes below.">
       Trim Reads:</span></td>
     <td><input type="checkbox" id="trimreads" name="trimreads" checked="checked" value="Yes"/></td>
    <tr>
    <tr id="tpadding" style="display:none">
     <td><span class="help"
       title="Specifiy the number of bases for option target padding. See Description and Notes below.">
       Target Padding:</span></td>
     <td><input type="text" size=6 id="padtargets" name="padtargets" value=0></select></td> </tr>
    <tr id="uniquereads">
     <td><span class="help"
       title="Check to analyze only reads mapped to a specific sequence in the genome. See Description and Usage Notes below.">
       Use Only Uniquely Mapped Reads:</span></td>
     <td><input type="checkbox" id="uniquemaps" name="uniquemaps" value="Yes" checked="checked"/></td> </tr>
    <tr id="nondups" style="display:none">
     <td><span class="help"
       title="Check to analyze only reads that are not demarked as duplicate reads. See Description and Usage Notes below.">
       Use Only Non-duplicate Reads:</span></td>
     <td><input type="checkbox" id="nonduplicates" name="nonduplicates" value="Yes" checked="checked"/>
     &nbsp;(Requires alignments with Mark Duplicates applied.)</td> </tr>
   </table>
   <br/>
   <div align="center" id="json_result">
    <input id="postbutton" type="submit" value="Submit">
   </div>
   <input type="hidden" id="librarytype_id" name="librarytype_id"/>
   <input type="hidden" id="targetregions_id" name="targetregions_id"/>
   <br/>
   <div align="left">
    <h3>Description and Usage Notes</h3>
    <p>
This plugin generates statistics, downloadable data and interactive visualization of coverage over
targeted regions of a reference.</p>
    <p>
The targeted regions selected should correspond to your reference and enriched fragment library.
Coverage will only be reported over the specified target regions.
Typically for Whole Genome library runs no target regions is specified.</p>
    <p>
Some options are only presented depending on the Library Type selected.
The Trim Reads option is only available for Ion AmpliSeq and is a recommended option. 
This specifies to initially trim all aligned reads to the overlaps with their assigned amplicon targets.
This avoids considering base reads that from primer sequence (or overlapping proximal targets) and
result in base coverage statistics that reflect those employed for variant calling (assuming that the
Uniquely Mapped Reads option is also checked).</p>
    <p>
The Target Padding value is only available for Ion TargetSeq.
This specifies a number of bases to expand each individual target by at both 3' and 5' ends and is 
applied to the selected target regions before coverage analysis is performed. A setting more than 0 is
not recommended for standard coverage analysis but a value of 100 or 150 may be useful for determining
the degree of near-target base reads.</p>
    <p>
The 'Uniquely Mapped Reads' and 'Non-duplicate Reads' options are used to specify coverage analysis on a filtered
subset of reads. Non-uniquely mapped reads are those that could have equally well aligned to multiple locations
in the whole genome and are typically mapped randomly to one of these locations. Hence, such reads are typically
no use for unambiguously locating variants and may result in a lower estimation of coverage for enriched libraries.
For whole genome libraries these reads may still be an accurate estimation of coverage over repeated regions.
Non-duplicate reads are those not marked as duplicates in the aligned reads (BAM) file and therefore the option
to use these is only effective if duplicate reads were so marked during the alignment process. This option is not
available for Ion AmpliSeq since (PCR) duplicate reads are anticipated and a necessary requirement for coverage.</p>
   </div>
 </form>
</body>
</html>

