<html>
<head>
 <script type="text/javascript" src="/site_media/jquery/js/jquery-1.6.1.min.js"></script>

 <style type="text/css">
  #formwrap {
   line-height: 2em;
   background: #eef;
   margin: 10px;
   padding: 10px;
   height: 500px;
   text-align: center;
  }
  body {
   font-family: "Lucida Grande", "Lucida Sans Unicode", Verdana, Arial, Helvetica, sans-serif;
   font-size: 14px;
  }
  .help {cursor:help; border-bottom: 1px dotted #A9A9A9}
 </style>

 <script>
  $(function()
  {
    // turn off async JavaScript so we can easily get/set var names
    $.ajaxSetup({async: false});
    // get the reference for the current report - TS_result is defined in run verison of this script
    var reportUrl = '/rundb/api/v1/results/' + TB_result;
    var referenceID;
    $.get( reportUrl, function(data)
    {
     var rl = data.log;
     var reportParams = rl.replace("log.html","ion_params_00.json");
     $.getJSON( reportParams, function(data)
     {
      // from report data get the short reference name
      referenceID = data.libraryName;
     });
    });
    document.getElementById('referenceid').innerHTML = referenceID;
    var refBedURL = '/rundb/api/v1/content/?format=json&publisher__name=BED&path__startswith=/' + referenceID + '/merged/plain/';
    $.get( refBedURL, function(data)
    {
      $.each( data.objects, function(intIndex,result)
      {
       var targfile = result.file;
       var i = targfile.lastIndexOf('/merged/plain/');
       if( i < 0 ) return true;
       var selName = targfile.substr(i+14)
       if( selName.substr(0,7) == 'example' ) return true;
       var j = selName.lastIndexOf('.bed');
       if( j > 0 && !result.meta.hotspot )
       {
        selName = selName.substr(0,j);
        $("#unptargets").append("<option value='" + targfile + "'>" +  selName + "</option>");
       }
      });
    });

    $.fn.serializeObject = function()
    {
     var o = {};
     var a = this.serializeArray();
     $.each(a,
      function() {
       if (o[this.name] != null) {
        if (!o[this.name].push) {
         o[this.name] = [o[this.name]];
        }
        o[this.name].push(this.value || '');
       } else {
        o[this.name] = this.value || '';
       }
      });
     return o;
    };

    $(function()
    {
     $('#postbutton').click(function()
     {
      var tarn = document.coverage.unptargets.selectedIndex;
      var regex = /^[0-9]+$/;
      var padval = document.coverage.padtargets.value;
      if( !regex.test(padval) )
      {
       alert("Target padding value must be a non-negative integer (>=0).");
       return false;
      }
      if( tarn == 0 && padval > 0 )
      {
       alert("Target padding value must be 0 if no specific Targeted regions value is selected.");
       return false;
      }
      var libn = document.coverage.librarytype.selectedIndex;
      var libType = document.getElementById('librarytype')[libn].value;
      var libName = document.getElementById('librarytype')[libn].text;
      if( libType == "fullgenome" )
      {
       if( tarn > 0 )
       {
        var msg = "Selecting targeted regions is typically not appropriate for Library Type: " + libName;
        msg += ". (Coverage will be considered in these just regions.)";
        msg += "\n\nDo you wish to continue?";
        if( !confirm(msg) ) return false;
       }
      }
      else
      {
       if( tarn == 0 )
       {
        var msg = "Targeted regions selection is typically appropriate for Library Type: " + libName;
        msg += ". (Variants may be reported in non-enriched regions where coverage is low.)";
        msg += "\n\nDo you wish to continue?";
        if( !confirm(msg) ) return false;
       }
      }
      var ustarts = document.coverage.uniquestarts.checked;
      if( libType != "targetseq" && ustarts == true )
      {
       var msg = "The Examine unique starts option will be ignored for Library Type: " + libName;
       msg += ".\n\nDo you wish to continue?";
       if( !confirm(msg) ) return false;
      }
      // replace spaces with underscores to avoid bug in TB
      var usrSel = document.getElementById('librarytype')[libn].text;
      document.coverage.librarytype_id.value = usrSel.replace(/\s/g,'_');
      usrSel = document.getElementById('unptargets')[tarn].text;
      document.coverage.targetregions_id.value = usrSel.replace(/\s/g,'_');
      obj = $('#coverage').serializeObject();
      pluginAPIJSON = { "plugin" : [TB_plugin.fields.name], "pluginconfig" : obj };
      pluginAPIJSON = JSON.stringify(pluginAPIJSON);
      pluginURL = "/rundb/api/v1/results/" + TB_result + "/plugin/";
      $.ajax({
       type: 'POST',
       url: pluginURL,
       async: false,
       contentType: "application/json; charset=utf-8",
       success: function(data)
       {
        $("#json_result").html('<div style="text-align: center;"><img src="/site_media/jquery/colorbox/images/loading.gif" alt="Running Plugin" style="float:center"></img><p>Running the Plugin... Check coverageAnalysis.html in the Plugin Summary</p></div>');
        setTimeout("parent.$.fn.colorbox.close()",2000);
       },
       data: pluginAPIJSON,
       dataType: "json"
      });
     });
    });

  });
 </script>
</head>

<body>
 <div align="center">
  <form id="coverage" name="coverage" align="center">
   <h1>Torrent Coverage Analysis Plugin</h1>
   <h3>Genome and Targeted Re-sequencing Coverage Analysis</h3>
   <table cellpadding="5px">
    <tr>
     <td><span class="help" title="The short name of the reference genome (or DNA sequences) that the current report was generated against and to be used for coverage analysis.">Reference Genome:</span></td>
     <td><div id="referenceid"></div></td>
    </tr>
    <tr>
     <td><span class="help" title="Select the library (enrichment) type sequenced.">Library Type:</span></td>
     <td><select id="librarytype" name="librarytype">
      <option value="fullgenome">Full Genome</option>
      <option value="ampliseq">Ion AmpliSeq</option>
      <option value="targetseq">Ion TargetSeq</option>
     </select></td>
    </tr>
    <tr>
    <tr>
     <td><span class="help"
       title="Select the target regions (ROI) matching your reference and enriched fragment library. See Description and Notes below.">
       Targeted regions:</span></td>
     <td><select id="unptargets" name="unptargets"><option value="">None</option></select></td> </tr>
    <tr>
     <td><span class="help"
       title="Specifiy the number of bases for on/off padded targets analysis. See Description and Notes below.">
       Target padding (bases):</span></td>
     <td><input type="text" size=6 id="padtargets" name="padtargets" value=0></select></td> </tr>
    <tr>
     <td><span class="help"
       title="Check to generate and examine coverage across reads filtered to unique starts. See Description and Notes below.">
       Examine unique starts:</span></td>
     <td><input type="checkbox" id="uniquestarts" name="uniquestarts" value="ustarts"/></td> </tr>
   </table>
   <br/>
   <div id="json_result">
    <input id="postbutton" type="submit" value="Submit">
   </div>
   <input type="hidden" id="librarytype_id" name="librarytype_id"/>
   <input type="hidden" id="targetregions_id" name="targetregions_id"/>
   <br/>
   <div align="left">
    <h3>Description and Usage Notes</h3>
    <p>
     This plugin generates statistics, downloadable data and basic visualization of coverage over targeted regions of a
     reference.</p>
    <p>
     The targeted regions selected should correspond to your reference and enriched fragment library. Coverage will only
     be reported over the specified target region.</p>
    <p>
     The Target padding value specifies a number of bases to expand each individual target by at both 3' and 5' ends
     for analyzing padded target coverage. The recommended value to use for Ion TargetSeq is 100.
     Padded targets coverage is only analyzed for on-target reads and results in
     one additional plot and summary statistic in the detailed coverage report(s). This data is not generated if the
     target padding value is set to 0 or the selected library type is Full Genome.</p>
    <p>
     The 'Examine unique starts' option is used to specify coverage analysis for a filtered subset of reads, generated by
     sampling a single uniquely mapped read per starting alignment to the reference, in both read orientations.
     This removes any bias due to non-uniform library enrichment but produces the most conservative representation
     of target coverage. When this option is checked, the detailed report table will have two columns of side-by-side
     results for the unfiltered (all) and the filtered (unique starts) reads. Unique starts coverage analysis is
     only performed if the selected library type is Ion TargetSeq.
    <p>
     A run will fail almost immediately (after queuing) if your selected targeted regions do not appear to match
     the reference that the reads were mapped to. (Check the log file if the plugin status is 'Error').</p>
   </div>
  </form>
  </div>
</body>
</html>

