<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE HTML>
<html>
<head>
 <script type="text/javascript" src="/site_media/jquery/js/jquery-1.6.1.min.js"></script>

 <style type="text/css">
  #formwrap {
   line-height: 2em;
   background: #eef;
   margin: 10px;
   padding: 10px;
   height: 500px;
  }
  body {
   font-family: "Lucida Grande", "Lucida Sans Unicode", Verdana, Arial, Helvetica, sans-serif;
   font-size: 14px;
  }
  .center { margin-left:auto; margin-right:auto; }
  .help {cursor:help; border-bottom: 1px dotted #A9A9A9}
 </style>

 <script>
  $(function() {

    function populateLibType(selectorID) {
      if( selectorID == '' ) return;
      $('#'+selectorID).empty();
      var runtypeUrl = '/rundb/api/v1/runtype/?format=json';
      $.get( runtypeUrl, function(data) {
        $.each( data.objects, function(intIndex,result) {
          var runType = result.runType;
          var selName = result.description.trim();
          // debar unsuported applications here, e.g. runType != "GENS"
          if( selName != "" ) {
            $('#'+selectorID).append("<option value='"+runType+"'>"+selName+"</option>");
          }
        });
      });
      orderSelectorList(selectorID);
    }

    function orderSelectorList(selectorID,firstItem) {
      var lst = $('#'+selectorID+' option');
      lst.sort(function(a,b) {
        if( a.text > b.text ) return 1;
        if( a.text < b.text ) return -1;
        return 0;
      });
      $('#'+selectorID).empty();
      if( firstItem ) {
        $('#'+selectorID).append("<option value=''>"+firstItem+"</option>");
      }
      $('#'+selectorID).append(lst);
      if( $('#'+selectorID+' option').size() > 0 )
        $('#'+selectorID+' option')[0].selected = true;
    }

    function setOptionDefaults(libType) {
      $("#libtypesel").val(libType);
      var showchecks = 'NYYY';
      if( libType == "AMPS_RNA" ) {
        showchecks = 'YNNN';
      } else if( libType.substr(0,4) == "AMPS" ) {
        showchecks = 'YNYN';
      } else if( libType == "TARS_16S" || libType == "RNA" ) {
        showchecks = 'NNNN';
      }
      $("#sampid").toggle(showchecks[0]=='Y');
      $("#tpadding").toggle(showchecks[1]=='Y');
      $("#uniquereads").toggle(showchecks[2]=='Y');
      $("#nondups").toggle(showchecks[3]=='Y');
    }

    $("#libtypesel").change(function() {
      var libType = $("#libtypesel option:selected").val();
      setOptionDefaults(libType);
    });

    // uncheck barcoded targets & sampleID options (allows values to be set consisently for checked state)
    $('#sampleid').attr('checked',false);
    $('#barcodebeds').attr('checked',false);
    $('#uniquemaps').attr('checked',false);
    $('.bcbed-shy').hide();
    
    // populate Library Type selector
    populateLibType("libtypesel");

    // populate Plan default selections
    var pluginLibType = "GENS";
    setOptionDefaults(pluginLibType);

  });
 </script>
</head>

<body>
 <form id="coverageAnalysis" name="coverageAnalysis">
  <div style="text-align:center">
   <h1>Torrent Coverage Analysis Plugin</h1>
   <h3>Genome and Targeted Re-sequencing Coverage Analysis</h3>
  </div>
   <table class="center" cellpadding="5px">
    <tr>
     <td><span class="help" title="Select the library (enrichment) type sequenced.">Library Type:</span></td>
     <td><select id="libtypesel"></select></td> </tr>
    <tr id="sampid" style="display:none">
     <td><span class="help"
       title="Check if this AmpliSeq library employed sampleID tracking amplicons.">
       Sample Tracking:</span></td>
     <td><input type="checkbox" id="sampleid" name="sampleid" checked="checked" value="Yes"/></td> </tr>
    <tr id="tpadding" style="display:none">
     <td><span class="help"
       title="Specifiy the number of bases for option target padding. See Description and Notes below.">
       Target Padding:</span></td>
     <td><input type="text" size=6 id="padtargets" name="padtargets" value=0></select></td> </tr>
    <tr id="uniquereads">
     <td><span class="help"
       title="Check to analyze only reads mapped to a specific sequence in the genome. See Description and Usage Notes below.">
       Use Only Uniquely Mapped Reads:</span></td>
     <td><input type="checkbox" id="uniquemaps" name="uniquemaps" value="Yes" checked="checked"/></td> </tr>
    <tr id="nondups" style="display:none">
     <td><span class="help"
       title="Check to analyze only reads that are not demarked as duplicate reads. See Description and Usage Notes below.">
       Use Only Non-duplicate Reads:</span></td>
     <td><input type="checkbox" id="nonduplicates" name="nonduplicates" value="Yes" checked="checked"/>
     &nbsp;(Requires alignments with Mark Duplicates applied.)</td> </tr>
   </table>
   <input type="hidden" id="launch_mode" name="launch_mode" value="Autostart with plan plugin configuration"/>
   <br/>
   <div align="left">
    <h3>Description and Usage Notes</h3>
    <p>
This plugin generates statistics, downloadable data files and interactive visualization of coverage over
targeted regions of the reference genome.</p>
    <p>
The Library Type should be set to the type of enriched fragment library used for the run (aka 'Run Type').
This is provided so that you may specify the options below that are dependent on the selection made.
The actual Library Type used at (auto) run time will be overriden by that specified in the Run Plan.
If this is different any non-applicable selections made will be ignorred in favor of the apropriate default settings.
    <p>
The SampleID Tracking option is recommended if your AmpliSeq panel was 'spiked' with SampleID sample tracking targets.
This will inform the analysis to produce an extra statistic for reads mapped to these targets so that
the true level of 'off-target' reads is accurately represented.</p>
    <p>
The Target Padding value is only available for the library types reporting with respect to targetted base coverage.
This specifies a number of bases to expand each individual target by at both 3' and 5' ends and is
applied to the selected target regions before coverage analysis is performed. A setting more than 0 is
not recommended for standard coverage analysis but a value of 100 or 150 may be useful for determining
the degree of near-target base reads.</p>
    <p>
The 'Uniquely Mapped Reads' and 'Non-duplicate Reads' options are used to specify coverage analysis on a filtered
subset of reads. Non-uniquely mapped reads are those that could have equally well aligned to multiple locations
in the whole genome and are typically mapped randomly to one of these locations. Such reads are typically
no use for unambiguously locating variants and may result in a lower estimation of coverage for enriched libraries.
For whole genome libraries these reads may still be an accurate estimation of coverage over repeated regions.
Non-duplicate reads are those not marked as duplicates in the aligned reads (BAM) file and therefore the option
to use these is only effective if duplicate reads were so marked during the alignment process. This option is not
available for Ion AmpliSeq since (PCR) duplicate reads are anticipated and a necessary requirement for coverage.</p>
   </div>
 </form>
</body>
</html>

